### PR TITLE
Re-enable -Werror on Unix on develop3

### LINKED ISSUE #3577: Turn back -Werror on Unix on develop3
https://github.com/jmarrec/OpenStudio/blame/d56398b5a89b2e95c98a12af4267a668545ca9c5/CMakeLists.txt#L342

### PR DESCRIPTION
**Pull request by jmarrec**
Fix #3577

I'm building it locally on Ubuntu, but plan to make use of CI to find mac errors.

### CODE REVIEW

**Comment by macumber** [comment]
@jmarrec is this done?  CI looks like it is still having trouble on Mac

**Comment by jmarrec** [comment]
@macumber I don't understand the mac error really, I put it on hold for now since I was getting nowhere with it. I haven't been able to create a MCVE for this (my MCVE works fine....)

Tried an MCVE, as far as I can tell this is the same as the stuff that's throwing an error, but this compiles fine on the Mac Mini used for CI even with -Werror :

```
#include <iostream>
#include <sstream>
#include <string>
#include <set>
#include <memory>

class IdfObject_Impl {
 public:
  IdfObject_Impl() {};
};

class IdfObject {
  public:
    IdfObject(std::shared_ptr<IdfObject_Impl> impl) : m_impl(std::move(impl)) {};
  
    template<typename T>
    std::shared_ptr<T> getImpl() const {
      return std::dynamic_pointer_cast<T>(m_impl);
    }
  
  private:
    // pointer to impl
    std::shared_ptr<IdfObject_Impl> m_impl;
};

struct IdfObjectImplLess {
  /** Function object for sorting by impl. */
  bool operator()(const IdfObject& left, const IdfObject& right) const {
    return (left.getImpl<IdfObject_Impl>() < right.getImpl<IdfObject_Impl>());
  };
  
};

int main()
{
  std::shared_ptr<IdfObject_Impl> B_impl( new IdfObject_Impl() );
  std::shared_ptr<IdfObject_Impl> A_impl( new IdfObject_Impl() );
  std::shared_ptr<IdfObject_Impl> C_impl;

  IdfObject A(A_impl);
  IdfObject B(B_impl);
  IdfObject B_bis(B_impl);
  std::set<IdfObject, IdfObjectImplLess> s;
  s.insert(A);
  s.insert(B);
  s.insert(B_bis);
  
  // shared_pt can be compared just fine: 
  std::cout << std::boolalpha << "A_impl < B_impl = " << (A_impl < B_impl) << std::endl;
  std::cout << std::boolalpha << "C_impl is nullptr = " << (C_impl == nullptr) << std::endl;

  std::cout << std::boolalpha << "A_impl < C_impl = " << (A_impl < C_impl) << std::endl;
  std::cout << std::boolalpha << "C_impl < A_impl = " << (C_impl < A_impl) << std::endl;


  std::cout << std::boolalpha << "Set size is 2: " << (s.size() == 2) << std::endl;
}
```

**Comment by jmarrec** [comment]
Jenkins is still throwing on mac, cf https://ci.commercialbuildings.dev/blue/organizations/jenkins/openstudio-incremental-develop3/detail/PR-3587/30/pipeline/8/#step-89-log-717:

```
In file included from /Users/jenkins/git/OpenStudioIncremental/develop3/src/gbxml/ForwardTranslator.cpp:30:
In file included from /Users/jenkins/git/OpenStudioIncremental/develop3/src/gbxml/ForwardTranslator.hpp:37:
In file included from /Users/jenkins/git/OpenStudioIncremental/develop3/src/gbxml/../model/../utilities/core/Logger.hpp:37:
In file included from /Users/jenkins/git/OpenStudioIncremental/develop3/src/gbxml/../utilities/core/Compare.hpp:41:
In file included from /Users/jenkins/.conan/data/boost_cycle_group_a/1.69.0/bincrafters/stable/package/641514453a9f65f8e0c7dfc14fa6b881970a8add/algorithm/include/boost/algorithm/string.hpp:23:
In file included from /Users/jenkins/.conan/data/boost_cycle_group_a/1.69.0/bincrafters/stable/package/641514453a9f65f8e0c7dfc14fa6b881970a8add/algorithm/include/boost/algorithm/string/split.hpp:16:
In file included from /Users/jenkins/.conan/data/boost_cycle_group_a/1.69.0/bincrafters/stable/package/641514453a9f65f8e0c7dfc14fa6b881970a8add/algorithm/include/boost/algorithm/string/iter_find.hpp:27:
In file included from /Users/jenkins/.conan/data/boost_cycle_group_a/1.69.0/bincrafters/stable/package/641514453a9f65f8e0c7dfc14fa6b881970a8add/algorithm/include/boost/algorithm/string/find_iterator.hpp:24:
In file included from /Users/jenkins/.conan/data/boost_cycle_group_a/1.69.0/bincrafters/stable/package/641514453a9f65f8e0c7dfc14fa6b881970a8add/algorithm/include/boost/algorithm/string/detail/find_iterator.hpp:18:
In file included from /Users/jenkins/.conan/data/boost_function/1.69.0/bincrafters/stable/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/function/include/boost/function.hpp:30:
In file included from /Users/jenkins/.conan/data/boost_function/1.69.0/bincrafters/stable/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/function/include/boost/function/detail/prologue.hpp:17:
In file included from /Users/jenkins/.conan/data/boost_function/1.69.0/bincrafters/stable/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/function/include/boost/function/function_base.hpp:21:
In file included from /Users/jenkins/.conan/data/boost_type_index/1.69.0/bincrafters/stable/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/type_index/include/boost/type_index.hpp:29:
In file included from /Users/jenkins/.conan/data/boost_type_index/1.69.0/bincrafters/stable/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/type_index/include/boost/type_index/stl_type_index.hpp:47:
In file included from /Users/jenkins/.conan/data/boost_container_hash/1.69.0/bincrafters/stable/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/container_hash/include/boost/container_hash/hash.hpp:760:
In file included from /Users/jenkins/.conan/data/boost_container_hash/1.69.0/bincrafters/stable/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/container_hash/include/boost/container_hash/extensions.hpp:22:
In file included from /Users/jenkins/.conan/data/boost_detail/1.69.0/bincrafters/stable/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/detail/include/boost/detail/container_fwd.hpp:95:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/set:426:26: error: the specified comparator type does not provide a const call operator [-Werror,-Wuser-defined-warnings]
    static_assert(sizeof(__diagnose_non_const_comparator<_Key, _Compare>()), "");
                         ^
/Users/jenkins/git/OpenStudioIncremental/develop3/src/gbxml/ForwardTranslator.hpp:110:74: note: in instantiation of template class 'std::__1::set<openstudio::model::Material, openstudio::IdfObjectImplLess, std::__1::allocator<openstudio::model::Material> >' requested here
    std::set<openstudio::model::Material, openstudio::IdfObjectImplLess> m_materials;
                                                                         
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/__tree:967:5: note: from 'diagnose_if' attribute on '__diagnose_non_const_comparator<openstudio::model::Material, openstudio::IdfObjectImplLess>':
    _LIBCPP_DIAGNOSE_WARNING(!std::__invokable<_Compare const&, _Tp const&, _Tp const&>::value,
    ^                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/__config:1251:21: note: expanded from macro '_LIBCPP_DIAGNOSE_WARNING'
     __attribute__((diagnose_if(__VA_ARGS__, "warning")))
                    ^           ~~~~~~~~~~~
1 error generated.
make[2]: *** [src/gbxml/CMakeFiles/openstudio_gbxml.dir/ForwardTranslator.cpp.o] Error 1
make[1]: *** [src/gbxml/CMakeFiles/openstudio_gbxml.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```

----

ForwardTranslator.hpp:

https://github.com/NREL/OpenStudio/blob/develop3/src/gbxml/ForwardTranslator.hpp#L108

https://github.com/NREL/OpenStudio/blob/develop3/src/gbxml/ForwardTranslator.hpp#L108

**Comment by jmarrec** [comment]
I've tried to augment my MCVE to try to get mac to throw the same error but to no avail. cf https://github.com/jmarrec/OpenStudio-test-mac-warning

**Comment by jmarrec** [comment]
Ok, at the point of declaration of std::set<Material, IdfObjectImplLess>, Material is just forward-declared so it's an incomplete type, and in particular the compiler can't know that Material derives from idfObject and will not correctly resolve the `operator()`.

**Comment by jmarrec** [comment]
@macumber I think this one is finally ready to drop in now!

File: CMakeLists.txt [hunk 1]
```diff
@@ -336,7 +339,7 @@ if(UNIX)
   #set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -fPIC -fno-strict-aliasing -Winvalid-pch -Wnon-virtual-dtor")
   set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS}  -fPIC -fno-strict-aliasing -Winvalid-pch")
   # Treat all warnings as errors
-  #set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Werror")
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Werror")
```
Comment (macumber) [inline CMakeLists.txt]: Hooray!

Comment (jmarrec) [reply CMakeLists.txt]: FYI CI is reporting failure but that's because of the failed unit tests (usual suspects). Everything built correctly etc.

**Comment by jmarrec** [comment]
@macumber can this drop in now?

**Pull request merged**